lib/segment/src/index/field_index/map_index.rs

```diff
@@ -175,6 +178,132 @@ impl<N: Hash + Eq + Clone + Display + FromStr> MapIndex<N> {
 
         Ok(())
     }
+
+    pub fn values_count(&self, point_id: PointOffsetType) -> usize {
+        self.get_values(point_id).map(|x| x.len()).unwrap_or(0)
+    }
+
+    /// Estimates cardinality for `except` clause
+    ///
+    /// # Arguments
+    ///
+    /// * 'excluded' - values, which are not considered as matching
+    ///
+    /// # Returns
+    ///
+    /// * `CardinalityEstimation` - estimation of cardinality
+    fn except_cardinality(&self, excluded: &[N]) -> CardinalityEstimation {
+        // Minimal case: we exclude as many points as possible.
+        // In this case, excluded points does not have any other values except excluded ones.
+        // So the first step - we estimate how many other points is needed to fit unused values.
+
+        // Example:
+        // Values: 20, 20
+        // Unique values: 5
+        // Total points: 100
+        // Total values: 110
+        // total_excluded_value_count = 40
+        // non_excluded_values_count = 110 - 40 = 70
+        // max_values_per_point = 5 - 2 = 3
+        // min_not_excluded_by_values = 70 / 3 = 24
+        // min = max(24, 100 - 40) = 60
+        // exp = ...
+        // max = min(20, 70) = 20
+
+        // Values, 60, 60
+        // Unique values: 5
+        // Total points: 100
+        // Total values: 200
+        // total_excluded_value_count = 120
+        // non_excluded_values_count = 200 - 120 = 80
+        // max_values_per_point = 5 - 2 = 3
+        // min_not_excluded_by_values = 80 / 3 = 27
+        // min = max(27, 100 - 120) = 27
+        // exp = ...
+        // max = min(60, 80) = 60
+
+        // Values, 60, 60, 60
+        // Unique values: 5
+        // Total points: 100
+        // Total values: 200
+        // total_excluded_value_count = 180
+        // non_excluded_values_count = 200 - 180 = 20
+        // max_values_per_point = 5 - 3 = 2
+        // min_not_excluded_by_values = 20 / 2 = 10
+        // min = max(10, 100 - 180) = 10
+        // exp = ...
+        // max = min(60, 20) = 20
+
+        let excluded_value_counts: Vec<_> = excluded
+            .iter()
+            .map(|val| self.map.get(val).map(|points| points.len()).unwrap_or(0))
+            .collect();
+        let total_excluded_value_count: usize = excluded_value_counts.iter().sum();
+
+        debug_assert!(total_excluded_value_count <= self.values_count);
+
+        let non_excluded_values_count =
+            self.values_count.saturating_sub(total_excluded_value_count);
+        let max_values_per_point = self.map.len().saturating_sub(excluded.len());
+
+        if max_values_per_point == 0 {
+            // All points are excluded, so we can't select any point
+            debug_assert_eq!(non_excluded_values_count, 0);
+            return CardinalityEstimation {
```

Comment:
A `CardinalityEstimation::zero()` would be nice here. I'll add that in a commit (after review).

Reply:
Suggested (in a different comment) to use `CardinalityEstimation::exact(0)` instead.